### PR TITLE
feat: allow excluding public channels from a legal hold

### LINKED ISSUE #55: Provide option where all public channels are excluded from legal hold
Some customers have requested this since those public channels are assumed to be accessible to everybody anyway, and do not need to be captured.  This will reduce disk space needed for those cases.

Related to https://github.com/mattermost/mattermost-plugin-legal-hold/issues/13

### PR DESCRIPTION
**Pull request by fmartingr**
#### Summary
Adds a new option to legal holds to avoid storing data from public channels

- [x] Adds `LegalHold.ExcludePublicChannels`
- [x] Modified signature of `GetChannelIDsForUserDuring` adding one parameter. (Probably better to refactor it using an `Opts` struct)
- [x] Modified the creation form to allow checking this option
- [x] Modified the edit form to allow checking this option (changes would only apply from the moment the change is made)
- [x] Better UI

#### Screenshots 
Screenshot 2024-06-06 at 16 13 37

Closes #55

### CODE REVIEW

File: server/store/sqlstore/legal_hold.go [hunk 1]
```diff
@@ -110,14 +110,24 @@ func (ss SQLStore) GetPostsBatch(channelID string, endTime int64, cursor model.L
...
 // a member of during the time period from (and including) the startTime up until (but not including) the
 // endTime.
-func (ss SQLStore) GetChannelIDsForUserDuring(userID string, startTime int64, endTime int64) ([]string, error) {
+func (ss SQLStore) GetChannelIDsForUserDuring(userID string, startTime int64, endTime int64, excludePublic bool) ([]string, error) {
 	query := ss.replicaBuilder.
 		Select("distinct(cmh.channelid)").
-		From("channelmemberhistory as cmh").
+		From("channelmemberhistory as cmh")
+
+	query = query.
```
Comment (mickmister) [inline server/store/sqlstore/legal_hold.go:118]: What's the improvement of this change?

**Review by mickmister — COMMENTED** [review]
Nice work @fmartingr. Do you mind adding some screenshots to the PR?

File: server/store/sqlstore/legal_hold.go (hunk 1 above)
Comment (fmartingr) [reply server/store/sqlstore/legal_hold.go:118]: That's just me making some tests. PR is in draft thus not ready to review. Will attach screenshots once finished, when the UI is somewhat pretty.

**Review by mickmister — APPROVED** [review]
LGTM

**Review by grundleborg — APPROVED** [review]

File: server/store/sqlstore/legal_hold.go [hunk 2]
```diff
@@ -110,14 +110,20 @@ func (ss SQLStore) GetPostsBatch(channelID string, endTime int64, cursor model.L
...
 		From("channelmemberhistory as cmh").
 		Where(sq.Lt{"cmh.jointime": endTime}).
 		Where(sq.Or{sq.Eq{"cmh.leavetime": nil}, sq.GtOrEq{"cmh.leavetime": startTime}}).
 		Where(sq.Eq{"cmh.userid": userID})
 
+	// Exclude all public channels from the results
+	if excludePublic {
+		query = query.Join("channels on cmh.channelid = channels.id").
+			Where(sq.NotEq{"channels.type": mattermostModel.ChannelTypeOpen})
+	}
```
[the remark this answers is not in the document]
Comment (fmartingr) [reply server/store/sqlstore/legal_hold.go:125]: Yeah, the original query didn't filter out anything.

**Comment by fmartingr** [comment]
@AayushChaudhary0001 can you add this to the QA pipeline please?

**Pull request merged**